src/Microsoft.TestPlatform.Client/TestPlatform.cs

```diff
@@ -211,81 +212,54 @@ public void ClearExtensions()
         }
     }
 
-    /// <summary>
-    /// Updates the test adapter paths provided through run settings to be used by the test
-    /// service.
-    /// </summary>
-    ///
-    /// <param name="runSettings">The run settings.</param>
-    private void AddExtensionAssemblies(string runSettings)
+
+    private void AddExtensionAssemblies(string runSettings, TestAdapterLoadingStrategy adapterLoadingStrategy)
     {
         IEnumerable<string> customTestAdaptersPaths = RunSettingsUtilities.GetTestAdaptersPaths(runSettings);
 
         if (customTestAdaptersPaths != null)
         {
             foreach (string customTestAdaptersPath in customTestAdaptersPaths)
             {
-                var adapterPath = Path.GetFullPath(Environment.ExpandEnvironmentVariables(customTestAdaptersPath));
-                if (!Directory.Exists(adapterPath))
-                {
-                    EqtTrace.Warning($"AdapterPath Not Found: {adapterPath}");
-                    continue;
-                }
+                var extensionAssemblies = ExpandTestAdapterPaths(customTestAdaptersPath, _fileHelper, adapterLoadingStrategy);
 
-                var extensionAssemblies = new List<string>(
-                    _fileHelper.EnumerateFiles(
-                        adapterPath,
-                        SearchOption.AllDirectories,
-                        TestPlatformConstants.TestAdapterEndsWithPattern,
-                        TestPlatformConstants.TestLoggerEndsWithPattern,
-                        TestPlatformConstants.DataCollectorEndsWithPattern,
-                        TestPlatformConstants.RunTimeEndsWithPattern));
-
-                if (extensionAssemblies.Count > 0)
+                if (extensionAssemblies.Any())
                 {
                     UpdateExtensions(extensionAssemblies, skipExtensionFilters: false);
                 }
+
             }
         }
     }
 
     /// <summary>
-    /// Updates the extension assemblies from source directory.
+    /// Updates the test logger paths from source directory.
     /// </summary>
     ///
-    /// <param name="testRunCriteria">The test run criteria.</param>
-    private void AddExtensionAssembliesFromSource(TestRunCriteria testRunCriteria)
+    /// <param name="sources">The list of sources.</param>
+    private void AddExtensionAssembliesFromSource(IEnumerable<string> sources, TestAdapterLoadingStrategy strategy)
     {
-        IEnumerable<string> sources = testRunCriteria.Sources;
-        if (testRunCriteria.HasSpecificTests)
+        // Skip discovery unless we're using the default behavior, or NextToSource is specified.
+        if (strategy != TestAdapterLoadingStrategy.Default 
+            && (strategy & TestAdapterLoadingStrategy.NextToSource) != TestAdapterLoadingStrategy.NextToSource)
```

Comment:
I don't know if you are using this syntax for the perf impact it used to have? I usually find the `HasFlag` function easier to understand.

Reply:
Old habits die hard :) I'll fix it.

Reply:
It's ok to keep it you prefer. I think the perf fix was done only in new framework.